staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go

```diff
@@ -301,25 +313,48 @@ type jsonPatcher struct {
 	fieldManager *fieldmanager.FieldManager
 }
 
-func (p *jsonPatcher) applyPatchToCurrentObject(currentObject runtime.Object) (runtime.Object, error) {
+func (p *jsonPatcher) applyPatchToCurrentObject(requestContext context.Context, currentObject runtime.Object) (runtime.Object, error) {
 	// Encode will convert & return a versioned object in JSON.
 	currentObjJS, err := runtime.Encode(p.codec, currentObject)
 	if err != nil {
 		return nil, err
 	}
 
 	// Apply the patch.
-	patchedObjJS, err := p.applyJSPatch(currentObjJS)
+	patchedObjJS, applyStrictJSONErrs, err := p.applyJSPatch(currentObjJS)
 	if err != nil {
 		return nil, err
 	}
 
 	// Construct the resulting typed, unversioned object.
 	objToUpdate := p.restPatcher.New()
 	if err := runtime.DecodeInto(p.codec, patchedObjJS, objToUpdate); err != nil {
-		return nil, errors.NewInvalid(schema.GroupKind{}, "", field.ErrorList{
-			field.Invalid(field.NewPath("patch"), string(patchedObjJS), err.Error()),
-		})
+		if !runtime.IsStrictDecodingError(err) && !runtime.IsPreservedDecodingError(err) {
```

Comment:
similar comment to https://github.com/kubernetes/kubernetes/pull/105916/files?w=1#r746107292 about expressing this more compactly (and avoiding the unreachable blocks)

Reply:
also, it looks like there are a couple bugs here:
* if `runtime.DecodeInto` returns no error, we ignore a non-empty applyStrictJSONErrs when we should be warning or failing
* if `runtime.DecodeInto` returns a non-strict error, we ignore a non-empty applyStrictJSONErrs when we should be warning

make sure there's test coverage for those scenarios, where the json patch has duplicate fields or unknown fields or something, and:
1. when applied, the resulting object decodes without error by DecodeInto (should get warned or fail with applyStrictJSONErrs)
2. when applied, the resulting object decodes with non-strict error by DecodeInto (should still get warned about applyStrictJSONErrs)